source/Nevermore/Advanced/WriteTransaction.cs

```diff
@@ -74,25 +83,60 @@ public class WriteTransaction : ReadTransaction, IRelationalTransaction
 
         public async Task InsertManyAsync<TDocument>(IReadOnlyCollection<TDocument> documents, InsertOptions options, CancellationToken cancellationToken = default) where TDocument : class
         {
-            IReadOnlyList<object> instanceList = documents.ToArray();
-            if (!instanceList.Any()) return;
+            IReadOnlyList<object> documentList = documents.ToArray();
+            if (!documentList.Any()) return;
+
+            var command = builder.PrepareInsert(documentList, options);
+
+            foreach (var document in documentList) await configuration.Hooks.BeforeInsertAsync(document, command.Mapping, this);
 
-            var command = builder.PrepareInsert(instanceList, options);
-            foreach (var instance in instanceList) await configuration.Hooks.BeforeInsertAsync(instance, command.Mapping, this);
             await ExecuteNonQueryAsync(command, cancellationToken);
-            foreach (var instance in instanceList) await configuration.Hooks.AfterInsertAsync(instance, command.Mapping, this);
-            configuration.RelatedDocumentStore.PopulateRelatedDocuments(this, instanceList);
+            foreach (var document in documentList) await RefreshRowVersionIfRequiredAsync(document, command);
+
+            foreach (var document in documentList) await configuration.Hooks.AfterInsertAsync(document, command.Mapping, this);
+
+            configuration.RelatedDocumentStore.PopulateRelatedDocuments(this, documentList);
         }
 
         public void Update<TDocument>(TDocument document, UpdateOptions options = null) where TDocument : class
         {
             var command = builder.PrepareUpdate(document, options);
             configuration.Hooks.BeforeUpdate(document, command.Mapping, this);
-            ExecuteNonQuery(command);
+            var numberOfUpdatedRows = ExecuteNonQuery(command);
+
+            AssertUpdateHasBeenSuccessful(numberOfUpdatedRows, command);
+            RefreshRowVersionIfRequired(document, command);
+
             configuration.Hooks.AfterUpdate(document, command.Mapping, this);
             configuration.RelatedDocumentStore.PopulateRelatedDocuments(this, document);
         }
 
+        void RefreshRowVersionIfRequired<TDocument>(TDocument document, PreparedCommand command) where TDocument : class
```

Comment:
Not sure if you want to make this optimization:
https://docs.microsoft.com/en-us/sql/t-sql/data-types/rowversion-transact-sql?view=sql-server-ver15

You can get the new RowVersion from the same `UPDATE` query (last code snippet)

Reply:
That would be nice. I don't see an easy way of incorporating this into NV right now. Any ideas?